### ISSUE TITLE
can not understand the loss of flatclr.py

### ISSUE THREAD

**Issue by yikeqingli** [Junya-Chen/FlatCLR#2]
Hi，Below line 114 of flatclr.py, no matter what the value of v is, the value of loss_vec will be a vector of all ones. Therefore, in line 119, loss_vec.mean()-1 must be 0. So what is the significance of this item? At the same time, detach is added after the cross_entropy of the item after loss. According to my understanding, this means that the calculated value of this item will not return the gradient. So how does the loss value optimize the network?

**Comment by Junya-Chen** [comment]
loss_vec.mean()-1 is equal to zero, but the gradient is not equal to zero. Thus, by minimizing the loss function by gradient descent, we can get a contrastive representation learning. cross_entropy.detach() is just used to show the progress of learning, which is not included in the actual loss function.

**Comment by ZifengLiu98** [comment]
Hi! Thanks for your great work and I have learned a lot from it. However I still have the confusion about the loss. 
I want to apply your flatclr in my supervised contrastive learning project. Since the value of loss is always 0, how can gradient descent to be completed? I added the entropy loss and the flatclr. It seems that the flatclr loss contribute nothing to the total loss.
Could u please help me to understand your flatclr? In my project it always equal to 0

**Comment by lxysl** [comment]
I am impressed by this brilliant work. But I have the same confusion about the code implementation. It seems the code implementation not completely match with the paper. I wonder if I got something wrong or lost.

In the paper, the pseudo code of the FlatNCE is as follow:
<img width="449" alt="image" src="https://user-images.githubusercontent.com/45818075/224205589-233232a0-227c-44ec-8fe6-84e5ca6a31d0.png">
In this repository, the implementation code of it is as follow:
```python
_, features = self.model(images)
logits, labels = self.flat_loss(features)
v = torch.logsumexp(logits, dim=1, keepdim=True) #(512,1)
loss_vec = torch.exp(v-v.detach())

assert loss_vec.shape == (len(logits),1)
dummy_logits = torch.cat([torch.zeros(logits.size(0),1).to(self.args.device), logits],1)
loss = loss_vec.mean()-1 + self.criterion(dummy_logits, labels).detach() #+
```
What confused me a lot is, isn't the `loss_vec` here is just the $l_{FlatNCE}$ in the pseudo code? What are the next a few lines doing? Can you help me with some explanation please?

Besides, I am working on integrate your FlatNCE loss to the [Supervised Contrastive Learning](https://github.com/HobbitLong/SupContrast). It's kind of difficult since there is a mask of the positive samples in it, instead of only one positive sample which is the sample itself. And also, there isn't a InfoNCE function or so in it. If you could also give me some guidance on it, I will be so much grateful!

**Comment by ZifengLiu98** [comment]
>


> I am impressed by this brilliant work. But I have the same confusion about the code implementation. It seems the code implementation not completely match with the paper. I wonder if I got something wrong or lost.
> 
> In the paper, the pseudo code of the FlatNCE is as follow: <img alt="image" width="449" src="https://user-images.githubusercontent.com/45818075/224205589-233232a0-227c-44ec-8fe6-84e5ca6a31d0.png"> In this repository, the implementation code of it is as follow:
> 
> ```python
> _, features = self.model(images)
> logits, labels = self.flat_loss(features)
> v = torch.logsumexp(logits, dim=1, keepdim=True) #(512,1)
> loss_vec = torch.exp(v-v.detach())
> 
> assert loss_vec.shape == (len(logits),1)
> dummy_logits = torch.cat([torch.zeros(logits.size(0),1).to(self.args.device), logits],1)
> loss = loss_vec.mean()-1 + self.criterion(dummy_logits, labels).detach() #+
> ```
> 
> What confused me a lot is, isn't the `loss_vec` here is just the lFlatNCE in the pseudo code? What are the next a few lines doing? Can you help me with some explanation please?
> 
> Besides, I am working on integrate your FlatNCE loss to the [Supervised Contrastive Learning](https://github.com/HobbitLong/SupContrast). It's kind of difficult since there is a mask of the positive samples in it, instead of only one positive sample which is the sample itself. And also, there isn't a InfoNCE function or so in it. If you could also give me some guidance on it, I will be so much grateful!

我在表情识别的任务中使用了flatclr,但是它的表现不及交叉熵损失，比[Supervised Contrastive Learning]提供的对比损失会稍微好点。我不确定是不是我使用有误，更大的batchsize下flatclr也没有更优

**Comment by Junya-Chen** [comment]
不好意思。我今天才看到邮件。我们确实使得loss是0，但是gradient并不是0。类似于这样的一个例子 loss =
x-x.detach()。loss会恒等于0，但是loss对x求导的话，由于第二项不参与求导，所以partial loss over partial
x = 1.
关于第二封邮件，也是同样的原因，loss=0很容易造成你的疑问，我们让loss看起来不等于0，所以我们补充了一些项，这些部分不会参与求导，但是可以帮助我们监控loss确实在下降，这样做只是为了看到一个我们习惯的loss。
我们觉得flatclr你作为baseline去对比的应该是infonce这个loss，supervised contrastive learning
已经用到了一部分supervised的信息了，我猜想你使用的交叉熵是纯supervised的结果。我之前看别的文献里面是讲supervised还是会比unsupervised好，如果你希望要用unsupervised，可以考虑只是作为预训练，最后还是要用supervised去做finetune，否则我自己的经验还是比不上的。关于大batchsize下，我们知道的是flatclr和infonce这样的contrastive
loss在batch size越大效果越好，当然需要的计算资源也越高。同样地，和supervised相比，可能还是有困难。

谢谢你对我们这个工作的关注。
我平时想不起来check这个邮箱，实在不好意思。你可以通过微信找到我:chenjunya5527,

junya

**Comment by Junya-Chen** [comment]
不好意思。我今天才看到邮件。我们确实使得loss是0，但是gradient并不是0。类似于这样的一个例子 loss =
x-x.detach()。loss会恒等于0，但是loss对x求导的话，由于第二项不参与求导，所以partial loss over partial
x = 1.
关于第二封邮件，也是同样的原因，loss=0很容易造成你的疑问，我们让loss看起来不等于0，所以我们补充了一些项，这些部分不会参与求导，但是可以帮助我们监控loss确实在下降，这样做只是为了看到一个我们习惯的loss。
我们觉得flatclr你作为baseline去对比的应该是infonce这个loss，supervised contrastive learning
已经用到了一部分supervised的信息了，我猜想你使用的交叉熵是纯supervised的结果。我之前看别的文献里面是讲supervised还是会比unsupervised好，如果你希望要用unsupervised，可以考虑只是作为预训练，最后还是要用supervised去做finetune，否则我自己的经验还是比不上的。关于大batchsize下，我们知道的是flatclr和infonce这样的contrastive
loss在batch size越大效果越好，当然需要的计算资源也越高。同样地，和supervised相比，可能还是有困难。

谢谢你对我们这个工作的关注。
junya

**Comment by LanXiaoPang613** [comment]
> Hi! Thanks for your great work and I have learned a lot from it. However I still have the confusion about the loss. I want to apply your flatclr in my supervised contrastive learning project. Since the value of loss is always 0, how can gradient descent to be completed? I added the entropy loss and the flatclr. It seems that the flatclr loss contribute nothing to the total loss. Could u please help me to understand your flatclr? In my project it always equal to 0

In my undestanding, the first term in the flatNce maintains zero, but the second terms will gererate different gradient during each iteration.
![image](https://github.com/Junya-Chen/FlatCLR/assets/35187545/27512efb-f699-4a07-a80b-4c63cdd60dd5)
